bentoml/_internal/bento/gen.py

```diff
@@ -115,71 +125,86 @@ def get_template_env(
+        use_conda (:code:`bool`, `required`):
+            Whether to use conda in the given Dockerfile.
 
+    Returns:
+        str: The rendered Dockerfile string.
 
-def generate_dockerfile(docker_options: DockerOptions, *, use_conda: bool) -> str:
-    distro = docker_options.distro
-    cuda_version = docker_options.cuda_version
-    user_templates = docker_options.dockerfile_template
+    .. dropdown:: Implementation Notes
 
-    templates_path = [os.path.join(os.path.dirname(__file__), "docker", "templates")]
-    if user_templates is not None:
-        dir_path = os.path.dirname(os.path.realpath(user_templates))
-        templates_path.append(dir_path)
+        The coresponding Dockerfile template will be determined automatically based on given :class:`DockerOptions`.
+        The templates are located `here <https://github.com/bentoml/BentoML/tree/main/bentoml/_internal/bento/docker/templates>`_
 
-    spec = DistroSpec.from_distro(
-        distro, cuda=cuda_version is not None, conda=use_conda
-    )
-    if spec is None:
-        raise BentoMLException("function is called before with_defaults() is invoked.")
+        As one can see, the Dockerfile templates are organized with the format :code:`<release_type>_<distro>.j2` with:
 
-    j2_template = f"{spec.release_type}_{distro}.j2"
+        +---------------+------------------------------------------+
+        | Release type  | Description                              |
+        +===============+==========================================+
+        | base          | A base setup for all supported distros.  |
+        +---------------+------------------------------------------+
+        | cuda          | CUDA-supported templates.                |
+        +---------------+------------------------------------------+
+        | miniconda     | Conda-supported templates.               |
+        +---------------+------------------------------------------+
+        | python        | Python releases.                         |
+        +---------------+------------------------------------------+
 
-    dockerfile_env = Environment(
-        extensions=["jinja2.ext.do", "jinja2.ext.loopcontrols", "jinja2.ext.debug"],
-        trim_blocks=True,
-        lstrip_blocks=True,
-        loader=FileSystemLoader(templates_path, followlinks=True),
-    )
-    dockerfile_env.globals.update(**J2_FUNCTION)  # type: ignore
+        Each of the templates will be validated correctly before being rendered. This will also include the user-defined custom templates
+        if :code:`dockerfile_template` is specified.
+
+    .. code-block:: python
+
+        from bentoml.build_utils import generate_dockerfile
+        from bentoml.build_utils import DockerOptions
 
-    validate_setup_blocks(dockerfile_env, j2_template)
-    bento_dockerfile_tmpl = dockerfile_env.get_template(j2_template)
+        dockerfile = generate_dockerfile(DockerOptions().with_defaults(), use_conda=False)
```

Comment:
docker_options = BentoInfo.from_yaml_file("{bento_path}/bento.yaml").docker
docker_options.docker_template = "./override_template.j2"
dockerfile = generate_dockerfile(docker_options, use_conda=False)

Reply:
update the docs here?